### PR TITLE
AIP-4600: [Part 2 of 3] Add in publishing to ZG Artifactory

### PR DESCRIPTION
**Pull request by alexlatchford**
Parts:
1. https://github.com/zillow/pipelines/pull/7
2. https://github.com/zillow/pipelines/pull/8 - This PR
3. https://github.com/zillow/pipelines/pull/9

This PR sets up building/publishing this repo as `zillow-kfp` to the internal ZG Artifactory on every commit to every branch; different versioning schemes exist for the `master` vs feature branches though.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+include:
+  - project: 'analytics/artificial-intelligence/ai-platform/aip-infrastructure/ci-templates/ci-cd-template'
+    ref: &include_ref 'alexla/AIP-4600'  # TODO AIP-4600 Replace with v2
```
Comment (hsezhiyan) [inline .gitlab-ci.yml:3]: `# TODO AIP-4600 Replace with v2` -> the end goal is to pin to `master`, right? Or would it be `feature/zg`?

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  extends: .aip_python_debian_image
+  stage: build
+  variables:
+    PY_LIBRARY_NAME: 'zillow-kfp'
+    KFP_VERSION_PATH: "sdk/python/kfp/__init__.py"
+  script:
+  - KFP_VERSION=$(cat $KFP_VERSION_PATH | sed -nr "s/^__version__ = ['\"]([^'\"]*)['\"]$/\1/p")
+  - |
+    if [ "${CI_COMMIT_BRANCH}" = "${CI_DEFAULT_BRANCH}" ]; then
+      PY_LIBRARY_VERSION="${KFP_VERSION}.${CI_PIPELINE_IID}"
```
Comment (hsezhiyan) [inline .gitlab-ci.yml:19]: Similar [comment](https://github.com/zillow/metaflow/pull/115#discussion_r704840171); would `KFP_VERSION` support semantic versioning?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    else
+      PY_LIBRARY_VERSION="0.0.${CI_PIPELINE_IID}+${CI_COMMIT_SHORT_SHA}"
+    fi
+  # Now write back the zillow-kfp version back to the original location we found the original.
+  - sed -i "s/\(__version__ = ['\"]\)[^'\"]*\(['\"]\)/\1${PY_LIBRARY_VERSION}\2/" $KFP_VERSION_PATH
```
Comment (hsezhiyan) [inline .gitlab-ci.yml:24]: I noticed we're duplicating this regex expression with the Metaflow publishing PR. Would it be worth storing in the `ci-cd templates` and giving it an easy to read name (e.g. `substitute_version_path_with_library_version`)?

Comment (alexlatchford) [reply .gitlab-ci.yml:24]: Yes and no. Whereas `kfp` exposes this as `__version__` in the `kfp/__init__.py`, `metaflow` exposes this as `version` in `setup.py` so I figured it was just different enough to not bother.

I think we likely need to spend some time making a paved path CICD template for libraries (trust me I wanted to do this in this round of PRs) but I decided given the scope of the current changes to not expand things to just move on with the `v3.9` upgrade 😅  Definitely something I'd like to investigate in the near future.

File: .gitlab-ci.yml (hunk 1 above)
Comment (alexlatchford) [reply .gitlab-ci.yml:3]: No the idea would be to replace the `ref` with `v2`. We now automatically `git tag` the `ci-cd-template` repos themselves when we make changes (see [this PR](https://gitlab.zgtools.net/analytics/artificial-intelligence/ai-platform/aip-infrastructure/ci-templates/ci-cd-template/-/merge_requests/105)). 

In [this PR](https://gitlab.zgtools.net/analytics/artificial-intelligence/ai-platform/aip-infrastructure/ci-templates/ci-cd-template/-/merge_requests/106) (that is coupled to the `v3.9` upgrade work), this bumps the version of the `ci-cd-template` publishing to `v2.0` and we use best effort tag overrides (not great but the least worst option Gitlab offers for any kind of semi-automated delivery mechanism, no such thing as lockfiles in Gitlab yet) to distribute updates to the CICD templates.

File: .gitlab-ci.yml (hunk 2 above)
Comment (alexlatchford) [reply .gitlab-ci.yml:19]: Yep, effectively the same policy as `zillow-metaflow`. In this case I'd expect the `zillow-kfp` version to be `1.7.0.{GITLAB_PIPELINE_ID}`.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
+include:
```
Comment (hsezhiyan) [inline .gitlab-ci.yml:1]: General question: if we add a `.gitlab-ci.yml` file at the root of this repo, wouldn't that complicate future OSS contributions? Wondering if you had some sort of mitigation strategy for that.

Comment (hsezhiyan) [inline .gitlab-ci.yml:1]: General question: if we add a `.gitlab-ci.yml` file at the root of this repo, wouldn't that complicate future OSS contributions? Wondering if you had some sort of mitigation strategy for that.

**Review by hsezhiyan — APPROVED** [review]

File: .gitlab-ci.yml (hunk 4 above)
Comment (alexlatchford) [reply .gitlab-ci.yml:1]: Yes and No. I think it depends on the commit strategy we have as likely we'd be making targeted contributions with a defined set of commits (ie. just the diffs we want to push) so I think we'd need to improve our `git` management strategy (for example when I've been committing to `kfp` I've been manually rebasing and squashing `git` commits so we push a very well known change set). 

So really if the diffs we push upstream don't include the `.gitlab-ci.yml` changes (or any other "internal" only for now/or forever) then in theory we can just keep those changes in our fork, wherever it lives in the repo.

**Pull request merged**